### PR TITLE
Do not throw exceptions from geometry classes

### PR DESCRIPTION
**Pull request by javagl**
Fixes https://github.com/CesiumGS/cesium-native/issues/205

As mentioned in the CHANGES.md:

> The `CesiumGeometry::Plane` and `CesiumGeometry::Ray` classes now offer factory
  methods to avoid throwing an `invalid_argument` exception in case of errors.
  They offer a `createUnchecked` method that assumes valid input, a `createOptional`
  method that returns an `std::optional` in case of errors, and a `createThrowing`
  method that throws an exception in case of errors. This has the effect of other
  classes also no longer throwing: The internal usage of these classes has been
  changed to generally use the `createUnchecked` method, because the preconditions
  for the exception to be thrown could not sensibly be propagated to the user.
  This means that the `CullingVolume` and `EllipsoidTangentPlane` do no longer
  throw for invalid inputs, but may return results that contain `NaN` values.

(Somewhat surprisingly, this does not seem to affect `cesium-unreal` on an API level: These classes had apparently not be used *directly*)

### CODE REVIEW

**Comment by javagl** [comment]
(An aside: Before merging, we should consider removing the `createOptional` and `createThrowing` methods: They are **not** used right now, and could be introduced later if they are required. The current state is mainly intended to show *how* they could be introduced later, but in a somewhat hopefully consistent manner)

**Comment by kring** [comment]
Thanks @javagl! The approach of Unchecked/Optional/Throwing makes sense to me. But methods like those should be a last resort when we can't avoid the issue entirely by changing the interface slightly. So with that in mind, let's talk about the three thrown exceptions that you've eliminated in this PR:

## EllipsoidTangentPlane::computeEastNorthUpToFixedFrame

This method would throw when the position is too close to the center of the ellipsoid. You've eliminated it by silently returning an identity matrix. That seems very reasonable to me, and is a good example of what I'm talking about above. When there's something sensible we can do to make an exceptional case non-exceptional, we should do it. The only thing I might suggest we change is to make the origin the position translation part of the matrix. The rotation/scale would still be identity.

## Plane constructor

It would previously throw if given a non-normalized normal. The old code looks like this:

```cpp
  //>>includeStart('debug', pragmas.debug);
  if (!Math::equalsEpsilon(glm::length(normal), 1.0, Math::EPSILON6)) {
    throw std::invalid_argument("normal must be normalized.");
  }
  //>>includeEnd('debug');
```

Those weird comments came from CesiumJS, and are the markers for code that gets removed in release builds. Therefore, a natural translation of that CesiumJS code to C++ is to use an assertion:

```cpp
assert(Math::equalsEpsilon(glm::length(normal), 1.0, Math::EPSILON6));
```

With that, we don't need exceptions or unchecked/optional/throwing methods. It's simply part of the contract that the normal must be normalized, and in debug builds we verify that this is true.

Why didn't I use an assertion to begin with? Uhh honestly this was some of the first code I wrote in cesium-native and I hadn't written C++ code in a little while so I had sort of forgotten that assertions exist. 😆

## Ray constructor

This case is exactly the same as the Plane constructor. It should be an assertion.

**Comment by javagl** [comment]
I just noticed that I didn't update the unit tests, but in hindsight, that's a good thing - I could have burnt some more time with that, considering that the suggested solution for the issue that caused this PR is now "just use `assert` instead of `throw`". Using `assert` *might* be some sort of "Best Practice" in C++, but I think it raises questions ...

- Do we really want to _always_ crash in these cases in Debug mode? 
- Do we really want to _never_ be informed about these cases in Release mode?
- What about the timing?

The latter refers to the problem that we don't exactly know where these cases appear. From ~"subjecive evidence", these crashes usually happened in some `Actor::Tick` call, that was apparently called at a point in time where some `dmat4` member had not (yet) been initialized properly (because some other function, `OnConstruction` or whatnot, had not been called yet). The complex interaction of things like `bCanEverTick`  and (particularly) `ShouldTickIfViewportsOnly` make it hard to analyze things like this *deterministically*. 

This is related to the difficulty of "propagating" statements about error conditions (aka preconditions) up the call hierarchy. This may be doable for specific, low-level cases like `computeEastNorthUpToFixedFrame`: *"Throws if position is too close to the center of the ellipsoid"*. How close? Where is this method called, and how will this condition be phrased at the call site?

    /* throws if ... some position is somewhere ... we don't know for sure ... better don't call this */
    void MyHighLevelBlueprintClass::update() {
        EllipsoidTangentPlane::computeEastNorthUpToFixedFrame(this->computePosition(), this->ellipsoid);
    }    

The general approach of "ignoring" invalid values may *appear* to be an easy solution: Sure, it will not crash any more. But if it then just does not render anything, and one has to figure out the reason for that, *without any clue what the reason might be* (only to find out that this is caused by some `NaN`, deeply hidden in some property), can literally involve *hours* of debug-stepping through Unreal Engine code. Having some `logVeryVerbose("Hey, the length should be x but is 0.0!")` could help to prevent that.

---

However: If this is *explicitly* reqested, I can create another PR, just replacing these checks with `assert`.

**Comment by kring** [comment]
> Do we really want to always crash in these cases in Debug mode?

Yes. Calling these methods with a normal that is not normalized is improper use of the methods. We should fail loudly. If you're in a situation where you're not sure whether a given normal will be normalized or not, the proper thing to do is to check before calling the method.

> Do we really want to never be informed about these cases in Release mode?

More like, by the time we're in a release build, we're assuming the library is being used properly and we don't need to spend cycles at runtime verifying this. Is it a perfect assumption? Of course not. I believe it's possible to configure a build to check assertions in a release build, if necessary, but it's rarely necessary.

> Having some logVeryVerbose("Hey, the length should be x but is 0.0!") could help to prevent that.

This is exactly the use-case that assertions are intended for: `assert(length == x)`. They're good for verifying preconditions that _should_ always be true but that the compiler / type system can't guarantee to always be true. Liberal use of assertions helps flag problems (violations of assumptions) early, rather than, say, letting a NaN propagate around until it causes an obvious problem later, by which point it's hard to tell where it came from.

> However: If this is explicitly reqested, I can create another PR, just replacing these checks with assert.

Just to be clear, this PR currently ignores these conditions in Ray and Plane entirely. What I'm proposing is that we instead assert the expected condition. I think what you've done with `computeEastNorthUpToFixedFrame` in this PR is correct. I wouldn't assert there.

**Comment by javagl** [comment]
I'm wondering about the "flow of events" for this. We did not encounter the reported crashes at development time (so we didn't trigger the exception, and thus, wouldn't have triggered the `assert`). In the wild, users either did something unexpected, or something unexpected happened for reasons beyond our knowledge (timing reasons or whatnot). In these cases, the exception was thrown. One problem was that it crashed the whole application, while some `log("Whoopsie in plane!")` would have been sufficient. But thanks to the stack trace, we knew **what** was wrong, exactly. **Iff** this had been an `assert` only, then ... people would have reported things like: "Hey, my stuff is not rendered" (or whatever happens when a the `NaN` spead through the UE). We wouldn't know what's wrong. We could only *hope* to be able to reproduce the problem in a debug build. That sounds like a potential debugging-time-sinkhole.

Long story short: I don't know the "silver bullet" here, but will doubtfully replace the `throw`s with `assert` in another PR.

**Comment by kring** [comment]
I don't disagree with you @javagl. But surely an assertion is better than the `createUnchecked` that was originally in this PR, right?

FWIW, I think the biggest problem was that the exception was causing immediate program termination in release builds due to the lack of exception support. The NaNs, while silent, may or may not be a serious problem, depending on how transient the condition that led to them is.

Anyway, exceptions are not an option in UE release builds. Even logging is questionable. So our options are extremely limited for this scenario where something goes wrong in a release build and we end up passing an unnormalized normal into this code.

**Comment by javagl** [comment]
> But surely an assertion is better than the `createUnchecked`

The way that I'm thinking about certain things revolves about the ability to _emulate_ one approach with the other. It's easier (and less of a "breaking change") to add an `assert` in a `createUnchecked` method. But having an `assert` in a constructor basides raising the questions that I mentioned) is a binding contract that can never be changed, and hardly even be modified, e.g. with adding a log statement. 

---

<sup>(I considered to move the following part to https://github.com/CesiumGS/cesium-native/issues/212 , but it focusses on `CesiumGeometry` in some way - there number of libraries in this repo is becoming too large...)</sup>

I tried to figure out some "best practices here", but ... that seems to be hard. I was surprised to see https://github.com/g-truc/glm/blob/master/manual.md#-710-why-some-glm-functions-can-crash-because-of-division-by-zero , for example: From my tests, it does *not* crash, but just create NaN. (Maybe whether or not it throws can be defined with some `#define`, but if it is, it is not properly documented). 

Other examples of ~"linear algebra"/~"geometry" libraries and their error handling could be 
http://eigen.tuxfamily.org/dox/TopicAssertions.html or https://doc.cgal.org/latest/Manual/devman_checks.html , but deriving a silver bullet from that may be hard as well: The latter plainly forbids `assert`, and they generally define own macros for checks with different granularities. Although macros are ... yuck... a relict from the 1970's ... , we might still consider this, if we make sure that they are cleanly defined, only used internally at a few critical places.

**Comment by kring** [comment]
> It's easier (and less of a "breaking change") to add an assert in a createUnchecked method. But having an assert in a constructor (besides raising the questions that I mentioned) is a binding contract that can never be changed, and hardly even be modified, e.g. with adding a log statement.

I don't think I follow you here. We're not concerned with breaking changes in cesium-native yet (other than we should try to document them). But even if we were, why is changing a constructor inherently more breaking than changing some other method?

**Comment by javagl** [comment]
> We're not concerned with breaking changes in cesium-native yet

Every time something changes, the calling code has to be reviewed, adapted, and tested extensively, in order to (only) decrease the likelihood of introducing a bug - even if it is "only" used in `cesium-native` itself, or in `cesium-unreal`. The "less breaking" refers to the fact that whatever the constructor does, it will always happen. If we want to offer anything that behaves differently (like a `createThrowing` method), then I think this will be much harder. (But maybe it's only a "technical inconvenience": it's tremendously hard to figure out where a constructor is called... except for ~"making it `private` and look through the compile errors" - instead of doing a full-text search for `Plane::createUnchecked`).

But I think that the general questions that are discussed here in this PR should be discussed in a broader sense (and led to a conclusion, as in a paradigm that is followed in `cesium-native` and preferably *all* other Cesium C++ libraries), so this should probably be done in https://github.com/CesiumGS/cesium-native/issues/212 (and in any case, this PR is obsolete, so I'll close it for now).
